### ISSUE TITLE
what should we export?

### ISSUE THREAD

**Issue by almartin82** [almartin82/mapvizieR#4]
hadley says:

> Generally, you want to export a minimal set of functions; the fewer you export, the 
> smaller the chance of a conflict.

we'll want to export most of our visualization functions - probably not much of an argument here.  I think that in general, whenever we have a wrapper around a series of workflow/prep tasks (eg `prep_cdf_long()`) then it is only necessary to export the final function,

we can make reference to the various helper functions in the package vignette.  

devil's advocate: would the existence of 'hidden' functions perhaps be a stumbling block to a new user?  are we being too clever by half?  like, whose namespace is really getting polluted by `cdf_prep_fws()`?

**Comment by chrishaid** [comment]
I think it depends on the user. For the advanced user, bit probably doesn't make a difference either way. You can pollute their environment and it won't be too big a deal; and if a function isn't exported they'll have the wherewithal to find it.

For the novice user, I think you follow Hadley's edict: less is better.  Especially if novice means the modal KIPP data person. Then we should make it as close to dead simple as possible for them. Take these simple steps and get pretty pics.  So I don't htink we need to pollute their namespace either.

So I guess, lets err on the side of less is more.  Further, I think if we think of the writing this for a novice, or with perhaps the opportunity to teach a course with this package, then we'll write more comprehensible code.  Rigth?

**Comment by almartin82** [comment]
Chris, any more thoughts on this?  If non-exported functions are hard to debug, I'm inclined to think that we should export just about everything.  Most of the hadley stack is in our supports/depends list anyway, so it's not like any of our functions are going to mess up someone's namespace with a conflict.  Having a new user (with some weird wrinkle to their data that we haven't thought of) reach an impasse that they can't `traceback()` out of is a real concern, but namespace conflicts feel real abstract.  :see_no_evil:?

**Comment by almartin82** [comment]
bumping this

hiding internals make things VERY hard to debug -- if you're stepping through the logic of a function line by line, and you hit a non-exported function, it returns an error.  i'm seeing this right now trying to pinpoint an error in haid_plot (not working in the package .Rproject, but rather working in an analytic workflow with our data).

that seals the argument for me -- given that all of our names tend to be domain specific or just weird (anyone else writing a `mv_limit_growth()` function?  bueller?) I say that we export everything by default.

**Comment by chrishaid** [comment]
Do you ever use `dplyr::dev_mode()` when developing?  That loads all the exported _and_ non exported functions in the package and gives you a clean(er?) namespace to develop in.

**Comment by almartin82** [comment]
For dplyr or for any package?

**Comment by almartin82** [comment]
[this](https://github.com/almartin82/mapvizieR/commit/83dc0d349c2391442132d21be8a94f5c2369dedd) commit is a perfect example - I needed to get at the internals of `haid_plot`, and doing so required triple prefixing some of the internal calls with `mapvizieR:::`.  which I promptly forgot to undo before pushing the patch back up.  

namespace _collisions_ are bad, but the risk of namespace collisions with these functions is infinitesimal.

**Comment by almartin82** [comment]
commit [#b1bc479](https://github.com/almartin82/mapvizieR/commit/b1bc47930cbe0a84d06293a9befa422df3326de7) exports `mv_opening_checks`, `mv_limit_growth`, etc

**Comment by chrishaid** [comment]
Shit. it's not `dplyr::dev_mode()`, rather it is `devtools:dev_mode`()`.  But that just gives you a clean environment.  The real trick is after running`dev_mode`is to run`devtools::load_all()`, the default behavior of which is to load all objects in the package _including non-exported functions_.  

It makes development and debugging easier, but there is a flip side, which is if you forget to export a function that you need to export you won't notice until you load the package later and try and export that function. 

Give a try!

**Comment by almartin82** [comment]
right, but what if you are _not_ on the package itself?  what if you have
simply loaded the package in another project (and have variables in that
environment that are causing a function to fail)?  that's the use case I
was solving for -- and none of the devtools stuff is much use in that
context, if I'm not mistaken
ᐧ

**Comment by chrishaid** [comment]
Ahh. I see. 

A couple of thoughts.
1. When we come accross them we should either fix them in our package or at least through more informative errors.
2.  I think for most users, exposing all those functions won't really help them.  To the extent that it does, then they would probably be going to the repo and using `package:::function` to get a sense fo what is going on, and then I think they'd submit a comment, and then we could recruit them to help improve the package. So what was a bug is now a feature!

**Comment by almartin82** [comment]
Closing as stale - this design discussion is 10 years old. The package's export strategy has been established through years of development and the current NAMESPACE reflects those decisions.
